File: tests/cuda/cpp/test_signal_filtering.cpp
// tests/cuda/test_signal_filtering.cpp
#include "insight/insight.h"
#include <cmath>
#include <complex>
#include <gtest/gtest.h>
#include <vector>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace ins;

namespace {

class SignalFilteringTestGPU : public ::testing::Test {
protected:
  static void SetUpTestSuite() {
    ins::init({"cpu", "iluvatar"});
    try {
      set_device(GPUPlace(0));
    // Iluvatar: signal API uses F64
    GTEST_SKIP() << "Iluvatar: signal API uses F64 (hardware limit)";
    } catch (...) {
      GTEST_SKIP() << "GPU not available";
    }
  }
};

template <typename T>
void expect_float_equal(const Array &arr, const std::vector<T> &expected,
                        T tol = 1e-6) {
  ASSERT_EQ(arr.numel(), static_cast<int64_t>(expected.size()));
  const T *data = arr.data<T>();
  for (int64_t i = 0; i < arr.numel(); ++i) {
    EXPECT_NEAR(data[i], expected[i], tol) << "mismatch at index " << i;
  }
}

// ========== hilbert ==========

TEST_F(SignalFilteringTestGPU, HilbertBasic) {
  int64_t N = 64;
  std::vector<double> x_data(N);
  for (int64_t i = 0; i < N; ++i) {
    x_data[i] = std::cos(2.0 * M_PI * i / N);
  }
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::hilbert(x, N);
  Array y_cpu = y.to(CPUPlace());
  ASSERT_EQ(y_cpu.dtype(), DType::C64);
  const std::complex<double> *data =
      reinterpret_cast<const std::complex<double> *>(y_cpu.data<char>());
  for (int64_t i = 0; i < N; ++i) {
    EXPECT_GT(std::abs(data[i]), 0.01);
  }
}

TEST_F(SignalFilteringTestGPU, HilbertDC) {
  std::vector<double> x_data = {1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0};
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::hilbert(x, 8);
  Array y_cpu = y.to(CPUPlace());
  const std::complex<double> *data =
      reinterpret_cast<const std::complex<double> *>(y_cpu.data<char>());
  for (int64_t i = 0; i < 8; ++i) {
    EXPECT_NEAR(data[i].real(), 1.0, 1e-10);
    EXPECT_NEAR(data[i].imag(), 0.0, 1e-10);
  }
}

// ========== detrend ==========

TEST_F(SignalFilteringTestGPU, DetrendConstant) {
  std::vector<double> x_data = {1.0, 2.0, 3.0, 4.0, 5.0};
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::detrend(x, -1, "constant");
  Array y_cpu = y.to(CPUPlace());
  const double *d = y_cpu.data<double>();
  EXPECT_NEAR(d[0], -2.0, 1e-10);
  EXPECT_NEAR(d[4], 2.0, 1e-10);
}

TEST_F(SignalFilteringTestGPU, DetrendLinear) {
  std::vector<double> x_data = {0.0, 3.0, 6.0, 9.0, 12.0};
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::detrend(x, -1, "linear");
  Array y_cpu = y.to(CPUPlace());
  const double *d = y_cpu.data<double>();
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], 0.0, 1e-10);
  }
}

// ========== lfilter (FIR) ==========

TEST_F(SignalFilteringTestGPU, LfilterFIRBasic) {
  std::vector<double> b_data = {1.0, 1.0};
  std::vector<double> a_data = {1.0};
  std::vector<double> x_data = {1.0, 0.0, 0.0, 0.0, 0.0};
  Array b = to_array(b_data);
  Array a = to_array(a_data);
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::lfilter(b, a, x);
  Array y_cpu = y.to(CPUPlace());
  EXPECT_EQ(y_cpu.numel(), 6);
  const double *d = y_cpu.data<double>();
  EXPECT_NEAR(d[0], 1.0, 1e-10);
  EXPECT_NEAR(d[1], 1.0, 1e-10);
}

// ========== lfilter (IIR) ==========

TEST_F(SignalFilteringTestGPU, LfilterIIRBasic) {
  std::vector<double> b_data = {1.0};
  std::vector<double> a_data = {1.0, -0.5};
  std::vector<double> x_data = {1.0, 0.0, 0.0, 0.0, 0.0};
  Array b = to_array(b_data);
  Array a = to_array(a_data);
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::lfilter(b, a, x);
  Array y_cpu = y.to(CPUPlace());
  const double *d = y_cpu.data<double>();
  EXPECT_NEAR(d[0], 1.0, 1e-10);
  EXPECT_NEAR(d[1], 0.5, 1e-10);
  EXPECT_NEAR(d[2], 0.25, 1e-10);
  EXPECT_NEAR(d[3], 0.125, 1e-10);
  EXPECT_NEAR(d[4], 0.0625, 1e-10);
}

// ========== lfilter_zi ==========

TEST_F(SignalFilteringTestGPU, LfilterZiIIR) {
  std::vector<double> b_data = {1.0};
  std::vector<double> a_data = {1.0, -0.5};
  Array b = to_array(b_data);
  Array a = to_array(a_data);
  Array zi = signal::lfilter_zi(b, a);
  EXPECT_EQ(zi.numel(), 1);
  EXPECT_NEAR(zi.data<double>()[0], 1.0, 1e-10);
}

// ========== filtfilt ==========

TEST_F(SignalFilteringTestGPU, FiltfiltFIR) {
  std::vector<double> b_data = {1.0, 1.0};
  std::vector<double> a_data = {1.0};
  std::vector<double> x_data = {1.0, 2.0, 3.0, 4.0, 5.0};
  Array b = to_array(b_data);
  Array a = to_array(a_data);
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::filtfilt(b, a, x);
  EXPECT_EQ(y.shape().dim(0), x.shape().dim(0));
}

// ========== freq_shift ==========

TEST_F(SignalFilteringTestGPU, FreqShiftBasic) {
  int64_t N = 64;
  std::vector<double> x_data(N, 1.0);
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::freq_shift(x, 0.0, 1.0);
  Array y_cpu = y.to(CPUPlace());
  ASSERT_EQ(y_cpu.dtype(), DType::C64);
  const std::complex<double> *data =
      reinterpret_cast<const std::complex<double> *>(y_cpu.data<char>());
  for (int64_t i = 0; i < N; ++i) {
    EXPECT_NEAR(std::abs(data[i]), 1.0, 1e-10);
  }
}

TEST_F(SignalFilteringTestGPU, FreqShiftPositive) {
  int64_t N = 16;
  std::vector<double> x_data(N, 1.0);
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::freq_shift(x, 0.25, 1.0);
  Array y_cpu = y.to(CPUPlace());
  const std::complex<double> *data =
      reinterpret_cast<const std::complex<double> *>(y_cpu.data<char>());
  for (int64_t i = 0; i < N; ++i) {
    double expected_real = std::cos(2.0 * M_PI * 0.25 * i);
    double expected_imag = std::sin(2.0 * M_PI * 0.25 * i);
    EXPECT_NEAR(data[i].real(), expected_real, 1e-10);
    EXPECT_NEAR(data[i].imag(), expected_imag, 1e-10);
  }
}

// ========== decimate ==========

TEST_F(SignalFilteringTestGPU, DecimateBasic) {
  int64_t N = 100;
  std::vector<double> x_data(N);
  for (int64_t i = 0; i < N; ++i) {
    x_data[i] = std::sin(2.0 * M_PI * i / N);
  }
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::decimate(x, 2, -1, false);
  EXPECT_EQ(y.numel(), 50);
}

// ========== resample ==========

TEST_F(SignalFilteringTestGPU, ResampleIdentity) {
  std::vector<double> x_data = {1.0, 2.0, 3.0, 4.0, 5.0};
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::resample(x, 5);
  Array y_cpu = y.to(CPUPlace());
  EXPECT_EQ(y_cpu.numel(), 5);
  const double *d = y_cpu.data<double>();
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], x_data[i], 1e-6);
  }
}

TEST_F(SignalFilteringTestGPU, ResampleUpsample) {
  std::vector<double> x_data = {1.0, 2.0, 3.0, 4.0};
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::resample(x, 8);
  EXPECT_EQ(y.numel(), 8);
}

// ========== resample_poly ==========

TEST_F(SignalFilteringTestGPU, ResamplePolyIdentity) {
  std::vector<double> x_data = {1.0, 2.0, 3.0, 4.0, 5.0};
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::resample_poly(x, 1, 1);
  Array y_cpu = y.to(CPUPlace());
  EXPECT_EQ(y_cpu.numel(), 5);
}

// ========== firfilter ==========

TEST_F(SignalFilteringTestGPU, FirfilterBasic) {
  std::vector<double> b_data = {1.0, 2.0, 1.0};
  std::vector<double> x_data = {1.0, 0.0, 0.0, 0.0, 0.0};
  Array b = to_array(b_data);
  Array x_cpu = to_array(x_data);
  Array x = x_cpu.to(GPUPlace(0));
  Array y = signal::firfilter(b, x);
  Array y_cpu = y.to(CPUPlace());
  EXPECT_EQ(y_cpu.numel(), 7);
  const double *d = y_cpu.data<double>();
  EXPECT_NEAR(d[0], 1.0, 1e-10);
  EXPECT_NEAR(d[1], 2.0, 1e-10);
  EXPECT_NEAR(d[2], 1.0, 1e-10);
}

} // namespace

File: tests/cuda/python/test_signal_filtering.py
"""Signal Filtering CUDA binding tests.

Run with:
    PYTHONPATH=bindings/python:build/bindings/python \
    LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
    python -m pytest tests/cuda/python/test_signal_filtering.py -v
"""

import sys
import os
import pytest
import math

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)
try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)


class TestHilbert:
    """Hilbert transform — tests 1-3."""

    def test_hilbert_basic(self):
        N = 64
        x_np = np.cos(2 * np.pi * np.arange(N) / N)
        x = ins.from_numpy(x_np.astype(np.float64))
        y = ins.signal.hilbert(x, N)
        assert y.numel == N
        # Analytic signal should have non-zero magnitude
        y_np = y.numpy()
        for i in range(N):
            assert abs(y_np[i]) > 0.01

    def test_hilbert_dc(self):
        x_np = np.ones(8, dtype=np.float64)
        x = ins.from_numpy(x_np)
        y = ins.signal.hilbert(x, 8)
        y_np = y.numpy()
        for i in range(8):
            assert abs(y_np[i].real - 1.0) < 1e-10
            assert abs(y_np[i].imag) < 1e-10

    def test_hilbert_default_n(self):
        x = ins.from_numpy(np.array([1.0, 2.0, 3.0, 4.0], dtype=np.float64))
        y = ins.signal.hilbert(x)
        assert y.numel == 4


class TestDetrend:
    """Detrend — tests 4-6."""

    def test_detrend_constant(self):
        x = ins.from_numpy(np.array([1.0, 2.0, 3.0, 4.0, 5.0], dtype=np.float64))
        y = ins.signal.detrend(x, type="constant")
        np.testing.assert_allclose(y.numpy(), [-2, -1, 0, 1, 2], atol=1e-10)

    def test_detrend_linear(self):
        x = ins.from_numpy(np.array([0.0, 3.0, 6.0, 9.0, 12.0], dtype=np.float64))
        y = ins.signal.detrend(x, type="linear")
        np.testing.assert_allclose(y.numpy(), 0.0, atol=1e-10)

    def test_detrend_linear_noisy(self):
        x = ins.from_numpy(np.array([0.1, 3.0, 5.9, 9.1, 12.0], dtype=np.float64))
        y = ins.signal.detrend(x, type="linear")
        np.testing.assert_allclose(y.numpy(), 0.0, atol=0.2)


class TestLfilter:
    """IIR/FIR filter — tests 7-9."""

    def test_lfilter_fir_basic(self):
        b = ins.from_numpy(np.array([1.0, 1.0], dtype=np.float64))
        a = ins.from_numpy(np.array([1.0], dtype=np.float64))
        x = ins.from_numpy(np.array([1, 0, 0, 0, 0], dtype=np.float64))
        y = ins.signal.lfilter(b, a, x)
        assert y.numel == 6
        d = y.numpy()
        assert abs(d[0] - 1.0) < 1e-10
        assert abs(d[1] - 1.0) < 1e-10
        for i in range(2, 6):
            assert abs(d[i]) < 1e-10

    def test_lfilter_fir_moving_avg(self):
        b = ins.from_numpy(np.array([1 / 3, 1 / 3, 1 / 3], dtype=np.float64))
        a = ins.from_numpy(np.array([1.0], dtype=np.float64))
        x = ins.from_numpy(np.array([1, 2, 3, 4, 5], dtype=np.float64))
        y = ins.signal.lfilter(b, a, x)
        d = y.numpy()
        assert abs(d[1] - 1.0) < 1e-10
        assert abs(d[2] - 2.0) < 1e-10
        assert abs(d[3] - 3.0) < 1e-10
        assert abs(d[4] - 4.0) < 1e-10

    def test_lfilter_iir_impulse(self):
        # y[n] = x[n] + 0.5*y[n-1], impulse response
        b = ins.from_numpy(np.array([1.0], dtype=np.float64))
        a = ins.from_numpy(np.array([1.0, -0.5], dtype=np.float64))
        x = ins.from_numpy(np.array([1, 0, 0, 0, 0], dtype=np.float64))
        y = ins.signal.lfilter(b, a, x)
        d = y.numpy()
        assert abs(d[0] - 1.0) < 1e-10
        assert abs(d[1] - 0.5) < 1e-10
        assert abs(d[2] - 0.25) < 1e-10
        assert abs(d[3] - 0.125) < 1e-10
        assert abs(d[4] - 0.0625) < 1e-10


class TestFiltfilt:
    """Zero-phase filtering — test 10."""

    def test_filtfilt_fir(self):
        b = ins.from_numpy(np.array([1.0, 1.0], dtype=np.float64))
        a = ins.from_numpy(np.array([1.0], dtype=np.float64))
        x = ins.from_numpy(np.array([1, 2, 3, 4, 5], dtype=np.float64))
        y = ins.signal.filtfilt(b, a, x)
        assert y.numel == x.numel


class TestFreqShift:
    """Frequency shift — tests 11-12."""

    def test_freq_shift_zero(self):
        N = 64
        x = ins.from_numpy(np.ones(N, dtype=np.float64))
        y = ins.signal.freq_shift(x, 0.0, 1.0)
        y_np = y.numpy()
        for i in range(N):
            assert abs(abs(y_np[i]) - 1.0) < 1e-10

    def test_freq_shift_positive(self):
        N = 16
        x = ins.from_numpy(np.ones(N, dtype=np.float64))
        fs = 1.0
        y = ins.signal.freq_shift(x, fs / 4.0, fs)
        y_np = y.numpy()
        for i in range(N):
            expected_real = math.cos(2 * math.pi * 0.25 * i)
            expected_imag = math.sin(2 * math.pi * 0.25 * i)
            assert abs(y_np[i].real - expected_real) < 1e-10
            assert abs(y_np[i].imag - expected_imag) < 1e-10


class TestDecimate:
    """Decimation — tests 13-14."""

    def test_decimate_basic(self):
        N = 100
        x_np = np.sin(2 * np.pi * np.arange(N) / N)
        x = ins.from_numpy(x_np.astype(np.float64))
        y = ins.signal.decimate(x, 2, zero_phase=False)
        assert y.numel == 50

    def test_decimate_zero_phase(self):
        N = 100
        x_np = np.sin(2 * np.pi * np.arange(N) / N)
        x = ins.from_numpy(x_np.astype(np.float64))
        y = ins.signal.decimate(x, 4, zero_phase=True)
        assert y.numel == 25


class TestResample:
    """Resampling — tests 15-16."""

    def test_resample_identity(self):
        x = ins.from_numpy(np.array([1, 2, 3, 4, 5], dtype=np.float64))
        y = ins.signal.resample(x, 5)
        assert y.numel == 5
        np.testing.assert_allclose(y.numpy(), [1, 2, 3, 4, 5], atol=1e-6)

    def test_resample_upsample(self):
        x = ins.from_numpy(np.array([1, 2, 3, 4], dtype=np.float64))
        y = ins.signal.resample(x, 8)
        assert y.numel == 8


class TestResamplePoly:
    """Polyphase resampling — tests 17-18."""

    def test_resample_poly_identity(self):
        x = ins.from_numpy(np.array([1, 2, 3, 4, 5], dtype=np.float64))
        y = ins.signal.resample_poly(x, 1, 1)
        assert y.numel == 5
        np.testing.assert_allclose(y.numpy(), [1, 2, 3, 4, 5], atol=1e-10)

    def test_resample_poly_upsample(self):
        x = ins.from_numpy(np.array([1, 0, 1, 0], dtype=np.float64))
        y = ins.signal.resample_poly(x, 2, 1)
        assert y.numel >= 7
        assert y.numel <= 9


class TestWiener:
    """Wiener filter — test 19."""

    def test_wiener_basic(self):
        # wiener requires 1D input (convolve only supports 1D tensors)
        x = ins.from_numpy(np.array([1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=np.float64))
        y = ins.signal.wiener(x)
        assert y.numel == 9


class TestFirfilter:
    """FIR filter — test 20."""

    def test_firfilter_basic(self):
        b = ins.from_numpy(np.array([1, 2, 1], dtype=np.float64))
        x = ins.from_numpy(np.array([1, 0, 0, 0, 0], dtype=np.float64))
        y = ins.signal.firfilter(b, x)
        assert y.numel == 7
        d = y.numpy()
        assert abs(d[0] - 1.0) < 1e-10
        assert abs(d[1] - 2.0) < 1e-10
        assert abs(d[2] - 1.0) < 1e-10
        for i in range(3, 7):
            assert abs(d[i]) < 1e-10


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
